vm/systemSmartContracts/auction_test.go

```diff
@@ -0,0 +1,510 @@
+		createABid(10000000, 10, 10000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+	}
+
+	expectedKeys := make([][]byte, 0)
+	for i, bid := range bids {
+		if i == 0 {
+			continue
+		}
+		expectedKeys = append(expectedKeys, bid.BlsPubKeys...)
+	}
+
+	data := stakingAuctionSC.selection(bids)
+	//check that 50% keys belong to the first bidder
+	count := 0
+	firstBidderKeys := bids[0].BlsPubKeys
+	for i, key := range data {
+		for j, expectedKey := range firstBidderKeys {
+			if bytes.Equal(key, expectedKey) {
+				firstBidderKeys = append(firstBidderKeys[:j], firstBidderKeys[j+1:]...)
+				data = append(data[:i], data[i+1:]...)
+				count++
+				break
+			}
+		}
+	}
+	assert.Equal(t, 5, count)
+	///////////////////////////////////////////////////////////
+
+	for _, key := range data {
+		for i, expectedKey := range expectedKeys {
+			if bytes.Equal(key, expectedKey) {
+				expectedKeys = append(expectedKeys[:i], expectedKeys[i+1:]...)
+				break
+			}
+
+			if i == len(expectedKeys)-1 {
+				assert.Equal(t, expectedKeys, data)
+				assert.Fail(t, "test fail")
+			}
+		}
+	}
+}
+
+func TestAuctionSC_selection_Case3PanicNumAllocatedNodesToBig(t *testing.T) {
+	t.Parallel()
+
+	minStakeValue := big.NewInt(1)
+	totalSupply := big.NewInt(100000000000)
+	minStep := big.NewInt(100000)
+	unBoundPeriod := uint64(100000)
+	numNodes := uint32(10)
+	eei := &mock.SystemEIStub{}
+	kg := &mock.KeyGenMock{}
+
+	stakingAuctionSC, _ := NewStakingAuctionSmartContract(minStakeValue, minStep, totalSupply, unBoundPeriod, numNodes, eei, kg)
+
+	bids := []AuctionData{
+		createABid(100000000, 10, 10000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+		createABid(1000000, 1, 1000000),
+	}
+
+	expectedKeys := make([][]byte, 0)
+	for i, bid := range bids {
+		if i == 0 {
+			continue
+		}
+		expectedKeys = append(expectedKeys, bid.BlsPubKeys...)
+	}
+
+	data := stakingAuctionSC.selection(bids)
+	//check that 50% keys belong to the first bidder
+	count := 0
+	firstBidderKeys := bids[0].BlsPubKeys
+	for i, key := range data {
+		for j, expectedKey := range firstBidderKeys {
+			if bytes.Equal(key, expectedKey) {
+				firstBidderKeys = append(firstBidderKeys[:j], firstBidderKeys[j+1:]...)
+				data = append(data[:i], data[i+1:]...)
+				count++
+				break
+			}
+		}
+	}
+	assert.Equal(t, 5, count)
+	///////////////////////////////////////////////////////////
+
+	for _, key := range data {
```

Comment:
duplicate code for this. Why not extract into a function and use it also on other tests?

Reply:
done

Reply:
done

Reply:
not done